File: 22-build-system/main.zig
const std = @import("std");

// ============================================================
// 演示：Zig 构建系统的核心概念
// ============================================================

// 获取编译目标信息（这是 Zig 独有的能力！）
const builtin = @import("builtin");

pub fn main() void {
    const print = std.debug.print;

    print("=== Zig 构建系统信息 ===\n\n", .{});

    // 编译期可以获取的目标平台信息
    print("目标架构: {s}\n", .{@tagName(builtin.cpu.arch)});
    print("目标操作系统: {s}\n", .{@tagName(builtin.os.tag)});
    print("目标 ABI: {s}\n", .{@tagName(builtin.abi)});
    print("字节序: {s}\n", .{@tagName(builtin.cpu.arch.endian())});
    print("指针宽度: {d} bits\n", .{@bitSizeOf(usize)});

    // 优化模式
    print("优化模式: {s}\n", .{@tagName(builtin.mode)});

    print("\n", .{});

    // 条件编译 - 基于目标平台
    print("=== 条件编译 ===\n\n", .{});

    if (builtin.os.tag == .linux) {
        print("运行在 Linux 上\n", .{});
    } else if (builtin.os.tag == .macos) {
        print("运行在 macOS 上\n", .{});
    } else if (builtin.os.tag == .windows) {
        print("运行在 Windows 上\n", .{});
    } else {
        print("运行在其他操作系统上\n", .{});
    }

    // 编译期计算版本信息
    const version = comptime getVersion();
    print("应用版本: {s}\n", .{version});

    // Zig 的优势：编译时确定的优化
    print("\n=== 构建模式说明 ===\n\n", .{});
    print("Zig 提供 4 种构建模式:\n", .{});
    print("  Debug        - 调试信息，安全检查，未优化\n", .{});
    print("  ReleaseSafe  - 优化 + 保留安全检查\n", .{});
    print("  ReleaseFast  - 最大性能优化\n", .{});
    print("  ReleaseSmall - 最小二进制体积\n", .{});

    print("\n=== 交叉编译 ===\n\n", .{});
    print("Zig 的交叉编译无需任何额外工具链:\n", .{});
    print("  zig build -Dtarget=aarch64-linux-gnu\n", .{});
    print("  zig build -Dtarget=x86_64-windows-gnu\n", .{});
    print("  zig build -Dtarget=riscv64-linux-gnu\n", .{});
    print("  zig build -Dtarget=wasm32-wasi\n", .{});
}

fn getVersion() []const u8 {
    return "1.0.0";
}

// ============================================================
// 演示：Zig 可以直接调用 C 代码（无需额外工具链）
// ============================================================

// 示例：使用 C 标准库（Zig 可以直接调用 C）
const c = @cImport({
    @cInclude("stdlib.h");
});

// 这个函数演示 Zig 与 C 的无缝互操作
pub fn cInteropDemo() void {
    // 调用 C 的 abs 函数
    const result = c.abs(-42);
    std.debug.print("C abs(-42) = {d}\n", .{result});
}

File: 22-build-system/main.rs
use std::env;

// ============================================================
// Rust 构建系统演示
//
// Rust 使用 Cargo 作为构建工具和包管理器。
//
// Cargo.toml 示例:
//   [package]
//   name = "my-app"
//   version = "1.0.0"
//   edition = "2021"
//
//   [dependencies]
//   serde = { version = "1.0", features = ["derive"] }
//   tokio = { version = "1", features = ["full"] }
//
//   [dev-dependencies]
//   criterion = "0.5"
//
//   [[bin]]
//   name = "my-app"
//   path = "src/main.rs"
//
//   [lib]
//   name = "my_lib"
//   path = "src/lib.rs"
//
//   [profile.release]
//   lto = true
//   codegen-units = 1
//   strip = true
//
// 常用命令:
//   cargo build              构建（debug 模式）
//   cargo build --release    构建（release 模式）
//   cargo run                编译并运行
//   cargo test               运行测试
//   cargo bench              运行基准测试
//   cargo doc --open         生成并打开文档
//   cargo clippy             代码 lint
//   cargo fmt                格式化代码
//   cargo add serde          添加依赖
//
// 交叉编译:
//   rustup target add aarch64-unknown-linux-gnu
//   cargo build --target aarch64-unknown-linux-gnu
//   # 通常还需要配置 .cargo/config.toml 指定链接器
//
// build.rs 示例（构建脚本）:
//   fn main() {
//       // 链接 C 库
//       println!("cargo:rustc-link-lib=sqlite3");
//
//       // 编译 C 源文件
//       cc::Build::new()
//           .file("src/helper.c")
//           .compile("helper");
//
//       // 生成代码
//       println!("cargo:rerun-if-changed=proto/api.proto");
//   }
// ============================================================

fn main() {
    println!("=== Rust 构建系统信息 ===\n");

    // 环境信息
    println!("目标架构: {}", env::consts::ARCH);
    println!("目标操作系统: {}", env::consts::OS);
    println!("目标系列: {}", env::consts::FAMILY);
    println!("可执行文件后缀: {:?}", env::consts::EXE_SUFFIX);
    println!("动态库后缀: {:?}", env::consts::DLL_SUFFIX);
    println!();

    // 条件编译
    println!("=== 条件编译 ===\n");

    #[cfg(target_os = "macos")]
    println!("运行在 macOS 上");

    #[cfg(target_os = "linux")]
    println!("运行在 Linux 上");

    #[cfg(target_os = "windows")]
    println!("运行在 Windows 上");

    #[cfg(target_arch = "x86_64")]
    println!("架构: x86_64");

    #[cfg(target_arch = "aarch64")]
    println!("架构: aarch64");

    #[cfg(debug_assertions)]
    println!("构建模式: Debug");

    #[cfg(not(debug_assertions))]
    println!("构建模式: Release");

    println!();

    // Cargo 命令说明
    println!("=== Cargo 常用命令 ===\n");
    println!("构建:");
    println!("  cargo build              # debug 模式（快速编译）");
    println!("  cargo build --release    # release 模式（优化）");
    println!("  cargo build --target ... # 交叉编译");
    println!();
    println!("运行与测试:");
    println!("  cargo run                # 编译并运行");
    println!("  cargo test               # 运行测试");
    println!("  cargo bench              # 基准测试");
    println!();
    println!("代码质量:");
    println!("  cargo clippy             # lint 检查");
    println!("  cargo fmt                # 代码格式化");
    println!("  cargo audit              # 安全审计");
    println!();

    // 与 Zig 对比的关键差异
    println!("=== 与 Zig 构建系统的关键差异 ===\n");
    println!("1. 交叉编译:");
    println!("   Zig:  zig build -Dtarget=aarch64-linux-gnu (开箱即用)");
    println!("   Rust: 需要 rustup target add + 配置链接器");
    println!();
    println!("2. C 互操作:");
    println!("   Zig:  内置 C 编译器，直接编译 C 源码");
    println!("   Rust: 需要 cc crate 和外部 C 编译器");
    println!();
    println!("3. 构建配置:");
    println!("   Zig:  build.zig (Zig 代码，完全可编程)");
    println!("   Rust: Cargo.toml (TOML 声明式) + build.rs (Rust 脚本)");
    println!();
    println!("4. 构建速度:");
    println!("   Zig:  编译速度很快（增量编译）");
    println!("   Rust: 首次编译较慢（但增量编译有改善）");
}
